server/datastore/mysql/software.go

```diff
@@ -2882,301 +3113,601 @@ last_vpp_install AS (
 	}
-	var hostSoftwareList []*hostSoftware
-	if err := sqlx.SelectContext(ctx, ds.reader(ctx), &hostSoftwareList, stmt, args...); err != nil {
-		return nil, nil, ctxerr.Wrap(ctx, err, "list host software")
+
+	var vppAdamIDs []string
+	for key := range byVPPAdamID {
+		vppAdamIDs = append(vppAdamIDs, key)
 	}
 
-	// collect the title ids to get the versions, vulnerabilities and installed
-	// paths for each software in the list.
-	titleIDs := make([]uint, 0, len(hostSoftwareList))
-	byTitleID := make(map[uint]*hostSoftware, len(hostSoftwareList))
-	for _, hs := range hostSoftwareList {
+	var titleCount uint
+	var hostSoftwareList []*hostSoftware
+	if len(softwareTitleIds) > 0 || len(vppAdamIDs) > 0 {
+		var args []interface{}
+		var stmt string
+		var softwareTitleStatment string
+		var vppAdamStatment string
+
+		matchClause := ""
+		matchArgs := []interface{}{}
+		if opts.ListOptions.MatchQuery != "" {
+			matchClause, matchArgs = searchLike(matchClause, matchArgs, opts.ListOptions.MatchQuery, "software_titles.name")
+		}
+
+		var softwareOnlySelfServiceClause string
+		var vppOnlySelfServiceClause string
+		if opts.SelfServiceOnly {
+			softwareOnlySelfServiceClause = ` AND software_installers.self_service = 1 `
+			if opts.IsMDMEnrolled {
+				vppOnlySelfServiceClause = ` OR ( vpp_apps_teams.self_service = 1 AND true ) `
+			} else {
+				vppOnlySelfServiceClause = ` OR ( vpp_apps_teams.self_service = 1 AND false ) `
+			}
+		}
 
-		// promote the package name and version to the proper destination fields
-		if hs.PackageName != nil {
-			var version string
-			if hs.PackageVersion != nil {
-				version = *hs.PackageVersion
+		var cveMetaFilter string
+		var hasCVEFilters bool
+		var cveMatchClause string
+		var cveNamedArgs []interface{}
+		var cveMatchArgs []interface{}
+		if opts.KnownExploit {
+			cveMetaFilter += "\nAND cve_meta.cisa_known_exploit = :known_exploit"
+			hasCVEFilters = true
+		}
+		if opts.MinimumCVSS > 0 {
+			cveMetaFilter += "\nAND cve_meta.cvss_score >= :min_cvss"
+			hasCVEFilters = true
+		}
+		if opts.MaximumCVSS > 0 {
+			cveMetaFilter += "\nAND cve_meta.cvss_score <= :max_cvss"
+			hasCVEFilters = true
+		}
+		if hasCVEFilters {
+			cveMetaFilter, cveNamedArgs, err = sqlx.Named(cveMetaFilter, namedArgs)
+			if err != nil {
+				return nil, nil, ctxerr.Wrap(ctx, err, "build named query for cve filters")
 			}
-			var platform string
-			if hs.PackagePlatform != nil {
-				platform = *hs.PackagePlatform
+		}
+		if opts.ListOptions.MatchQuery != "" {
+			cveMatchClause, cveMatchArgs = searchLike(cveMatchClause, cveMatchArgs, opts.ListOptions.MatchQuery, "software_cve.cve")
+		}
+
+		var softwareVulnerableJoin string
+		if len(softwareTitleIds) > 0 {
+			if opts.VulnerableOnly || opts.ListOptions.MatchQuery != "" {
+				softwareVulnerableJoin += " AND ( "
+				if !opts.VulnerableOnly && opts.ListOptions.MatchQuery != "" {
+					softwareVulnerableJoin += `
+						NOT EXISTS (
```

Comment:
Not following what's going on here. Seems like we're doing a "not exists or exists" when we could...just zero out the query when appropriate and have the existence check (or an inner join?) when we're filtering by vulnerable?

Reply:
This one is a bit tricky because with the previous logic if we aren't matching on "prefixed with cve-", we could technically be searching for a software name or a cve name. So this becomes an `OR`.
If we use an inner join, then we would filter out software that doesn't have any vulnerabilities but match the term. 

For example searching for `ve` returns vulnerable software with cve in the name or software with ve in the name that's not vulnerable.
<img width="1551" alt="Screenshot 2025-04-08 at 3 46 40 PM" src="https://github.com/user-attachments/assets/1fa82e6f-1d9a-4e33-9f03-ac4e6d735168" />

Not sure if that's a hard requirement, but causes me to do all these nasty `NOT IN` queries.

Reply:
https://github.com/fleetdm/fleet/issues/27003 <- present in the testing criteria. Also how our "software search" works.

Reply:
Deferring to @noahtalerman and @jmwatts on this, but I _think_ that "don't bother trying to search for vulnerable software if the search query isn't cve-something" gets us a more predictable UX, compared to "why the heck does searching for 've' turn up all of this software that doesn't have 've' in the nam...oh, well that makes things harder to find." See also: searching for "v" or "e" or "c". The non-host-specific Software page has this issue already of course ("why does "Finder" show up when I search for "v").

If there isn't a response in the next 24h or so, fair enough on keeping this as-is.

Reply:
I agree with @iansltx, I think it's pretty odd that when we are searching for a cve we can get software that isn't vulnerable in the results. I do appreciate the one box to search them all approach, however, I think it should be smart about the user's intent where possible. 

I think if we change the behavior, we should change it in both places. @noahtalerman, let me know if that's the behavior you would like and I can create a ticket to change it in both locations.

aka

"CVE-<[\d-]+>" <- this is the only way to search for vulnerable software and will only return vulnerable software. Otherwise we are just searching through software names.

Reply:
https://github.com/fleetdm/fleet/issues/28039